### PR TITLE
Allow duplicate file names in project hierarchy

### PR DESCRIPTION
**Pull request by robrankin**
The Project Factory module doesn't allow duplicate filenames in the project hierarchy folder structure, as the filenames are currently used as keys in TF maps, and duplicates lead to duplicate key errors.

This change moves to building the map keys from either:

* The full path + filename with the separator of `/` replaced with `-` (in the abscence of an explicit `name:` key in the yaml)
* The `name:` key explicitly set in the project yaml definition file.

I've included two example project files, both named `ce-0.yaml`, but in different parts of the project hierarchy.

The first file, `team-a/dev/ce-0.yaml` has an explicitly set `name:` key, and produces the following resource:
```
  # module.projects.module.projects["ta-d-ce-0"].google_project.project[0] will be created
  + resource "google_project" "project" {
      + name                = "main-ta-d-ce-0"
      + project_id          = "main-ta-d-ce-0"
    }
```

The second file, `team-b/dev/ce-0.yaml` does not have an explicitly set `name:` key, and produces the following resource with the project name, ID and TF map key built from the path + filename:
```
  # module.projects.module.projects["team-b-dev-ce-0"].google_project.project[0] will be created
  + resource "google_project" "project" {
      + name                = "main-team-b-dev-ce-0"
      + project_id          = "main-team-b-dev-ce-0"
    }
```

---
**Checklist**

If applicable, I acknowledge that I have:
- [X] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [X] Ran `terraform fmt` on all modified files
- [X] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [X] Made sure all relevant tests pass

### CODE REVIEW

**Comment by juliocc** [comment]
Not a bad idea but if we do this for PF we should probably do it for all other factories too. 

I actually like the idea and the fix is fairly simple.

@ludoo wdyt?

File: modules/project-factory/factory-projects.tf [hunk 1]
```diff
@@ -20,7 +20,7 @@ locals {
   _hierarchy_projects = (
     {
       for f in try(fileset(local._folders_path, "**/*.yaml"), []) :
-      basename(trimsuffix(f, ".yaml")) => merge(
+      replace(trimsuffix(f, ".yaml"), "/", "-") => merge(
```
Comment (juliocc) [inline modules/project-factory/factory-projects.tf:23]: do you really need the replace? I think you could keep '/' in the keys.

Comment (juliocc) [reply modules/project-factory/factory-projects.tf:23]: Well yes if you want to use the filename as project id. 

Ignore

Comment (ludoo) [reply modules/project-factory/factory-projects.tf:23]: Same comment I made above, path should be left intact

**Review by ludoo — COMMENTED** [review]
Forgot to send comments lol

File: modules/project-factory/factory-projects.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliocc) [reply modules/project-factory/factory-projects.tf:23]: the problem (I think) is that '/' is invalid in project ids

[the remark this answers is not in the document]
Comment (ludoo) [reply modules/project-factory/factory-projects.tf:23]: We need to keep the oath for keys, not project ids

**Comment by robrankin** [comment]
Anything I can do here?  Happy to make changes if needed.

**Comment by ludoo** [comment]
> Anything I can do here? Happy to make changes if needed.

Can you implement my suggestion above? Just remove basename and trim the suffix, and then we see how it works :)

**Comment by ludoo** [comment]
Rob I'm closing this, let's reopen a discussion to flesh out what we want to do here.

**Pull request closed without merging**